system/lib/wasmfs/wasmfs.cpp

```diff
@@ -6,63 +6,94 @@
 
 extern "C" {
 
-static std::vector<char> fd_write_stdstream_buffer;
+using namespace wasmfs;
+
+__wasi_fd_t __syscall_dup2(__wasi_fd_t oldfd, __wasi_fd_t newfd) {
+  FileTable::Handle fileTable = FileTable::get();
+
+  auto oldOpenFile = fileTable[oldfd];
+  // If oldfd is not a valid file descriptor, then the call fails,
+  // and newfd is not closed.
+  if (!oldOpenFile) {
+    return __WASI_ERRNO_BADF;
+  }
+
+  if (oldfd == newfd) {
+    return oldfd;
+  }
+
+  if (newfd < 0) {
+    return __WASI_ERRNO_BADF;
+  }
+
+  // If the file descriptor newfd was previously open, it is closed
+  // before being reused; the close is performed silently.
+  if (fileTable[newfd]) {
+    fileTable.remove(newfd);
+  }
+
+  fileTable[newfd] = oldOpenFile;
+  return newfd;
+}
+
+__wasi_fd_t __syscall_dup(__wasi_fd_t fd) {
+
+  FileTable::Handle fileTable = FileTable::get();
+  if (!fileTable[fd]) {
+    return __WASI_ERRNO_BADF;
+  }
+  // Find the first free open file entry
+  return fileTable.add(fileTable[fd]);
+}
 
 __wasi_errno_t __wasi_fd_write(
   __wasi_fd_t fd, const __wasi_ciovec_t* iovs, size_t iovs_len, __wasi_size_t* nwritten) {
-  // FD 1 = STDOUT and FD 2 = STDERR.
-  // Temporary hardcoding of filedescriptor values.
-  // TODO: May not want to proxy stderr (fd == 2) to the main thread.
-  // This will not show in HTML - a console.warn in a worker is suffficient.
-  // This would be a change from the current FS.
-  if (fd == 1 || fd == 2) {
-    __wasi_size_t num = 0;
-    for (size_t i = 0; i < iovs_len; i++) {
-      const uint8_t* buf = iovs[i].buf;
-      __wasi_size_t len = iovs[i].buf_len;
-      for (__wasi_size_t j = 0; j < len; j++) {
-        uint8_t current = buf[j];
-        if (current == 0 || current == 10) {
-          fd_write_stdstream_buffer.push_back('\0'); // for null-terminated C strings
-          if (fd == 1) {
-            emscripten_console_log(&fd_write_stdstream_buffer[0]);
-          } else if (fd == 2) {
-            emscripten_console_error(&fd_write_stdstream_buffer[0]);
-          }
-          fd_write_stdstream_buffer.clear();
-        } else {
-          fd_write_stdstream_buffer.push_back(current);
-        }
-      }
-      num += len;
-    }
-    *nwritten = num;
+  // Get the corresponding OpenFile from the open file table
+  FileTable::Handle fileTable = FileTable::get();
+  if (!fileTable[fd]) {
+    return __WASI_ERRNO_BADF;
   }
-  return 0;
+
+  __wasi_size_t num = 0;
+  for (size_t i = 0; i < iovs_len; i++) {
+    const uint8_t* buf = iovs[i].buf;
+    __wasi_size_t len = iovs[i].buf_len;
+
+    fileTable[fd]->get().getFile()->get().write(buf, len);
+    num += len;
+  }
+  *nwritten = num;
+
+  return __WASI_ERRNO_SUCCESS;
 }
 
 __wasi_errno_t __wasi_fd_seek(
   __wasi_fd_t fd, __wasi_filedelta_t offset, __wasi_whence_t whence, __wasi_filesize_t* newoffset) {
   emscripten_console_log("__wasi_fd_seek has been temporarily stubbed and is inert");
-  abort();
+  return __WASI_ERRNO_INVAL;
```

Comment:
I would keep the `abort()` in these cases because the function truly has not been implemented. In the case of `read` on stdin, for instance, returning the error code was the entire implementation.